src/commands/deleteVirtualMachine/deleteVirtualMachine.ts

```diff
@@ -0,0 +1,55 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import { ComputeManagementClient } from "@azure/arm-compute";
+import { DialogResponses, IActionContext, IAzureQuickPickItem } from "vscode-azureextensionui";
+import { ext } from "../../extensionVariables";
+import { localize } from "../../localize";
+import { VirtualMachineTreeItem } from "../../tree/VirtualMachineTreeItem";
+import { createComputeClient } from "../../utils/azureClients";
+import { nonNullValue } from "../../utils/nonNull";
+import { IDeleteChildImplContext, ResourceToDelete, virtualMachineLabel } from "./deleteConstants";
+import { getResourcesAssociatedToVm } from "./getResourcesAssociatedToVm";
+import { promptResourcesToDelete } from "./promptResourcesToDelete";
+
+export async function deleteVirtualMachine(context: IActionContext & Partial<IDeleteChildImplContext>, node?: VirtualMachineTreeItem): Promise<void> {
+    if (!node) {
+        node = await ext.tree.showTreeItemPicker<VirtualMachineTreeItem>(VirtualMachineTreeItem.allOSContextValue, context);
+    }
+
+    const resourcesP: Promise<ResourceToDelete[]> = new Promise<ResourceToDelete[]>(async (resolve, _reject): Promise<void> => {
```

Comment:
We want to avoid "new Promise" and "then" stuff when possible because it can lead to some tricky issues. For example, if any of your code within "new Promise" throws an exception before `resolve(associatedResources)`, this promise will never return and the command will fail without even prompting the user.

I would basically get rid of the `promptResourcesToDelete.ts` file and structure the code like this:
```typescript
const resourcesToDelete: IAzureQuickPickItem<ResourceToDelete>[] = await ext.ui.showQuickPick(getQuickPicks(node), { placeHolder: localize('selectResources', 'Select resources to delete'), canPickMany: true });
```

```typescript
async function getQuickPicks(node: VirtualMachineTreeItem): Promise<IAzureQuickPickItem<ResourceToDelete>[]> {
    const resources: ResourceToDelete[] = await getResourcesAssociatedToVm(node);

    // add the vm to the resources to delete since it is not an associated resource
    resources.unshift({
        resourceName: node.name, resourceType: virtualMachineLabel, picked: true,
        deleteMethod: async (): Promise<void> => {
            const computeClient: ComputeManagementClient = await createComputeClient(node.root);
            await computeClient.virtualMachines.deleteMethod(node.resourceGroup, node.name);
        }
    });

    return resources.map(resource => {
        return { label: resource.resourceName, description: resource.resourceType, data: resource, picked: resource.picked };
    });
}
```

If you keep `getQuickPicks` in this file, you can keep the name simple like that. If you put it in a separate file, you'd probably want a more unique name